### PR TITLE
Fix for caching appearing broken in batch2 for bulkexport jobs

### LINKED ISSUE #3911: Caching appears broken in Batch2 for BulkExport jobs.
Steps to Reproduce: 
1. Create a patient
2. Create a bulk export job for a single patient
3. Note that the location header is correctly populated with a job id. 
4. Perform step 2 again. 
5. Note that the content header is returning a _new_ ID and it should be cached
6. note that the returned binary is empty

### PR DESCRIPTION
**Pull request by tadgh**
* Propagate cache property to job starter. 
* If job exits early with no work chunks, set completed date. 
* Fix broken HQL queries
* Fix paging bug. 
* Test
* Changelog

Closes #3911

### CODE REVIEW

**Review by fil512 — APPROVED** [review]
NIce work!

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportProviderTest.java [hunk 1]
```diff
@@ -655,4 +655,41 @@ public void testProviderProcessesNoCacheHeader() throws IOException {
...
+		post.setEntity(new ResourceEntity(myCtx, input));
+		ourLog.info("Request: {}", post);
+		try (CloseableHttpResponse response = myClient.execute(post)) {
+			ourLog.info("Response: {}", response.toString());
+			assertEquals(202, response.getStatusLine().getStatusCode());
+			assertEquals("Accepted", response.getStatusLine().getReasonPhrase());
+			assertEquals("http://localhost:" + myPort + "/$export-poll-status?_jobId=" + theExpectedJobId, response.getFirstHeader(Constants.HEADER_CONTENT_LOCATION).getValue());
+		}
+	}
+
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportProviderTest.java:693]: NIce test.

**Review by epeartree — APPROVED** [review]

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java [hunk 2]
```diff
@@ -80,8 +82,10 @@ void executeStep() {
 		}
 
 		if (stepExecutorOutput.getDataSink().firstStepProducedNothing()) {
-			ourLog.info("First step of job myInstance {} produced no work chunks, marking as completed", myInstanceId);
-			myJobPersistence.markInstanceAsCompleted(myInstanceId);
+			ourLog.info("First step of job myInstance {} produced no work chunks, marking as completed and setting end date", myInstanceId);
+			myInstance.setEndTime(new Date());
+			myInstance.setStatus(StatusEnum.COMPLETED);
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java:87]: If completed statuses are expected to have an end-time, it feels like there should be a single method you call that does both of these things.  I worry there are other places where we set the job to completed but forget to set the end time

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java (hunk 2 above)
Comment (tadgh) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java:87]: Yeah you're likely correct, opening a ticket to address this in the future.

**Pull request merged**